bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java

```diff
@@ -807,7 +807,7 @@ private boolean switchToNewBufferOrSuspendRead(final byte[] array, final int off
 
     private boolean switchToNewBuffer(final byte[] array, final int offset, final int size)
     {
-      if (datalist.isMemoryBlockAvailable()) {
+      if (datalist.isMemoryBlockAvailable() || datalist.canAllocateMemoryBlock()) {
```

Comment:
It makes sense to move the storage == null check for the second part out into the caller. Will do that.